doc/design/containerd/port-layer-integration.md

```diff
@@ -0,0 +1,276 @@
+
+<TBD>
+
+**Questions**
+
+There are a number of questions that still need to be worked out. Here’s are the most obvious:
+
+*Do we need to create a process per container in the VIC appliance to fit with the executor model?*
+
+Both runc and the containerd shim are a process-per-container model. Our Port Layer is a long-running service. We will need some kind of shim layer on the Port Layer to mate with the executor interface anyway. Does having entries in /proc for each container help in any way?
+
+*What do we do with the config.json in a bundle?*
+
+When a bundle is extracted to a local filesystem, config.json is used as input to runc in terms of how to run the container. Config.json in the normal case will be a composite of image config and runtime config generated by containerd. For us to integrate seamlessly, we should expect the output of containerd to still be a config.json, but this will then need to be parsed by the Port Layer in creation of the VMConfigSpec. The config.json does not then need to end up in the bundle filesystem mounted to the container since runtime configuration goes into VM guestInfo.
+
+From discussions with the containerd team, it doesn't seem necessary to cache the config.json anywhere or make it available to the container. In fact, containerd is not responsible for modifying config.json at all.
+
+*What parts of config.json can/should we ignore?*
+
+VIC is not the only hypervisor OCI implementation out there and it should be noted that there are a bunch of areas in the OCI config spec that simply can’t apply to VIC. There is no distinction in the spec between the OS that supports the container infrastructure and the OS that the container runs on - one implies the other. In the case of VIC, the two are completely distinct. 
+
+Other hypervisor implementations list a bunch of areas that they either [don’t support or ignore](https://github.com/01org/cc-oci-runtime/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20label%3Adomain%3AOCI). The spec itself lists a bunch of parts that are "optional". We should put a proposal together in another document.
+
+*Role of the Supervisor component*
+
+The Supervisor is a multiplexer for shims, so it's very similar in scope and function to our Port Layer.
+
+*Are snapshots created during image pull or container create?*
+
+In the current model, snapshots are created on pull. Containerd seems to be designed to have snapshots created as part of container create. This needs to be clarified as it will have significant performance implications on create as well as caching implications for the image data.
+
+*What to do with pause and resume on the executor interface?*
+
+This could map to VM suspend and resume, but to date this is not something we’ve tried or even seen a need for. According to the containerd team, this is for the purpose of committing images. 
```

Comment:
We could use it to pause VM, it might be necessary if people decide to continue further investigation of incorrectly running container on some other VM deployment. Just saying, not sure if it is even possible.